addons/hr_holidays/models/res_company.py

```diff
@@ -0,0 +1,132 @@
+        holiday_end_utc = user_tz.localize(datetime.combine(holiday_date, time.max)).astimezone(pytz.utc).replace(tzinfo=None)
+        return holiday_start_utc, holiday_end_utc
+
+    def _generate_public_holidays(self, year_range, convert_datetime=True):
+        response = []
+        existing_holidays = self.env["resource.calendar.leaves"].search([
+            ('company_id', 'in', self.ids),
+            ('resource_id', '=', False),
+        ])
+        existing_holidays_ranges = [
+            (holiday.company_id, holiday.date_from, holiday.date_to)
+            for holiday in existing_holidays
+        ]
+        subdivision_per_country_dict = holidays.list_supported_countries(include_aliases=True)
+        lang_per_country = holidays.list_localized_countries(include_aliases=True)
+
+        for company in self:
+            if not company.country_code:
+                response.append({
+                    'title': self.env._('No Country Code'),
+                    'type': 'danger',
+                    'message': self.env._('Please select a country in %(company)s to load public holidays.', company=company.name),
+                })
+                continue
+            if company.country_code not in subdivision_per_country_dict:
+                response.append({
+                    'title': self.env._('No Public Holidays'),
+                    'type': 'danger',
+                    'message': self.env._('Public holidays are not available for %(country)s country.', country=company.country_id.name),
+                })
+                continue
+
+            company_subdiv = company.state_id.code
+            subdiv = subdivision_per_country_dict.get(company.country_code, None)
+            user_lang_iso_code = self.env["res.lang"]._lang_get(self.env.user.lang).iso_code[:2]
+            if not user_lang_iso_code or user_lang_iso_code not in lang_per_country.get(company.country_code, []):
+                user_lang_iso_code = 'en_US'
+
+            public_holiday_dict = holidays.country_holidays(
+                company.country_code,
+                subdiv=company_subdiv if subdiv and company_subdiv in subdiv else None,
+                years=year_range,
+                language=user_lang_iso_code,
+            )
+
+            overlapped_holidays = []
+            user_tz = pytz.timezone(company.resource_calendar_id.tz)
+            public_holidays_values_dict = {}
+
+            for holiday_date, holiday_name in public_holiday_dict.items():
+                holiday_start_utc, holiday_end_utc = self._convert_to_utc(holiday_date, user_tz)
+                overlapping = any(
+                    company_id in (company, False) and
+                    date_from < holiday_end_utc and
+                    date_to > holiday_start_utc
+                    for company_id, date_from, date_to in existing_holidays_ranges
+                )
```

Comment:
existing_holidays_ranges should be filtered for the company. Note that there might also be public holidays created manually for specific calendars as well, it shouldn't prevent to have a full-company public leave type to be created in my opinion :slightly_smiling_face: